File: backend/src/clients/sonarr.client.ts
import axios, { AxiosInstance, AxiosError } from 'axios';
import { config } from '../config/env';
import logger from '../utils/logger';

interface SonarrSeries {
  id: number;
  title: string;
  sortTitle: string;
  status: string;
  overview: string;
  network: string;
  airTime: string;
  images: Array<{
    coverType: string;
    url: string;
  }>;
  seasons: Array<{
    seasonNumber: number;
    monitored: boolean;
  }>;
  year: number;
  path: string;
  qualityProfileId: number;
  languageProfileId: number;
  seasonFolder: boolean;
  monitored: boolean;
  tvdbId: number;
  tvRageId: number;
  tvMazeId: number;
  imdbId: string;
  titleSlug: string;
  rootFolderPath: string;
  added: string;
}

interface SonarrSearchResult {
  title: string;
  sortTitle: string;
  status: string;
  overview: string;
  network: string;
  airTime: string;
  images: Array<{
    coverType: string;
    url: string;
  }>;
  seasons: Array<{
    seasonNumber: number;
    monitored: boolean;
  }>;
  year: number;
  tvdbId: number;
  tvRageId: number;
  tvMazeId: number;
  imdbId: string;
  titleSlug: string;
}

interface SonarrQueueItem {
  id: number;
  seriesId: number;
  episodeId: number;
  series: {
    title: string;
  };
  episode: {
    seasonNumber: number;
    episodeNumber: number;
    title: string;
  };
  quality: {
    quality: {
      name: string;
    };
  };
  size: number;
  title: string;
  sizeleft: number;
  timeleft: string;
  estimatedCompletionTime: string;
  status: string;
  trackedDownloadStatus: string;
  trackedDownloadState: string;
  downloadId: string;
  protocol: string;
  downloadClient: string;
  indexer: string;
  outputPath: string;
}

interface SonarrQueueResponse {
  page: number;
  pageSize: number;
  sortKey: string;
  sortDirection: string;
  totalRecords: number;
  records: SonarrQueueItem[];
}

interface AddSeriesOptions {
  tvdbId: number;
  title: string;
  qualityProfileId: number;
  rootFolderPath: string;
  seasonFolder?: boolean;
  monitored?: boolean;
  searchForMissingEpisodes?: boolean;
}

export class SonarrClient {
  private client: AxiosInstance;
  private baseURL: string;
  private apiKey: string;

  constructor(baseURL?: string, apiKey?: string) {
    this.baseURL = baseURL || config.externalServices.sonarr.url;
    this.apiKey = apiKey || config.externalServices.sonarr.apiKey;

    if (!this.apiKey) {
      logger.warn('Sonarr API key not configured');
    }

    this.client = axios.create({
      baseURL: this.baseURL,
      timeout: 30000,
      headers: {
        'X-Api-Key': this.apiKey,
        'Content-Type': 'application/json'
      }
    });

    // Response interceptor for logging
    this.client.interceptors.response.use(
      (response) => response,
      (error: AxiosError) => {
        logger.error('Sonarr API error', {
          url: error.config?.url,
          status: error.response?.status,
          message: error.message,
          data: error.response?.data
        });
        throw error;
      }
    );
  }

  /**
   * Update configuration dynamically
   */
  updateConfig(baseURL?: string, apiKey?: string): void {
    if (baseURL) this.baseURL = baseURL;
    if (apiKey) this.apiKey = apiKey;

    // Reinitialize the client with new config
    this.client = axios.create({
      baseURL: this.baseURL,
      timeout: 30000,
      headers: {
        'X-Api-Key': this.apiKey,
        'Content-Type': 'application/json'
      }
    });

    this.client.interceptors.response.use(
      (response) => response,
      (error: AxiosError) => {
        logger.error('Sonarr API error', {
          url: error.config?.url,
          status: error.response?.status,
          message: error.message,
          data: error.response?.data
        });
        throw error;
      }
    );

    logger.info('Sonarr configuration updated');
  }

  /**
   * Test connection to Sonarr
   */
  async testConnection(): Promise<boolean> {
    try {
      const response = await this.client.get('/api/v3/system/status');
      logger.info('Sonarr connection successful', { version: response.data.version });
      return true;
    } catch (error) {
      logger.error('Sonarr connection failed', { error });
      return false;
    }
  }

  /**
   * Get health check status
   */
  async getHealth(): Promise<any> {
    try {
      const response = await this.client.get('/api/v3/health');
      return response.data;
    } catch (error) {
      logger.error('Failed to get Sonarr health status', { error });
      throw error;
    }
  }

  /**
   * Search for series by title
   */
  async searchSeries(query: string): Promise<SonarrSearchResult[]> {
    try {
      const response = await this.client.get<SonarrSearchResult[]>('/api/v3/series/lookup', {
        params: { term: query }
      });
      return response.data;
    } catch (error) {
      logger.error('Failed to search series in Sonarr', { query, error });
      throw error;
    }
  }

  /**
   * Add a series to Sonarr
   */
  async addSeries(options: AddSeriesOptions): Promise<SonarrSeries> {
    try {
      const payload = {
        tvdbId: options.tvdbId,
        title: options.title,
        qualityProfileId: options.qualityProfileId,
        titleSlug: options.title.toLowerCase().replace(/[^a-z0-9]+/g, '-'),
        images: [],
        seasons: [],
        path: `${options.rootFolderPath}/${options.title}`,
        rootFolderPath: options.rootFolderPath,
        seasonFolder: options.seasonFolder !== false,
        monitored: options.monitored !== false,
        addOptions: {
          searchForMissingEpisodes: options.searchForMissingEpisodes !== false
        }
      };

      const response = await this.client.post<SonarrSeries>('/api/v3/series', payload);
      logger.info('Series added to Sonarr', { title: options.title, id: response.data.id });
      return response.data;
    } catch (error) {
      logger.error('Failed to add series to Sonarr', { options, error });
      throw error;
    }
  }

  /**
   * Get all series
   */
  async getSeries(): Promise<SonarrSeries[]> {
    try {
      const response = await this.client.get<SonarrSeries[]>('/api/v3/series');
      return response.data;
    } catch (error) {
      logger.error('Failed to get series from Sonarr', { error });
      throw error;
    }
  }

  /**
   * Get series by ID
   */
  async getSeriesById(id: number): Promise<SonarrSeries> {
    try {
      const response = await this.client.get<SonarrSeries>(`/api/v3/series/${id}`);
      return response.data;
    } catch (error) {
      logger.error('Failed to get series by ID from Sonarr', { id, error });
      throw error;
    }
  }

  /**
   * Get download queue
   */
  async getQueue(page = 1, pageSize = 20): Promise<SonarrQueueResponse> {
    try {
      const response = await this.client.get<SonarrQueueResponse>('/api/v3/queue', {
        params: {
          page,
          pageSize,
          includeUnknownSeriesItems: false
        }
      });
      return response.data;
    } catch (error) {
      logger.error('Failed to get queue from Sonarr', { error });
      throw error;
    }
  }

  /**
   * Delete series from Sonarr
   */
  async deleteSeries(id: number, deleteFiles = false): Promise<void> {
    try {
      await this.client.delete(`/api/v3/series/${id}`, {
        params: {
          deleteFiles,
          addImportListExclusion: false
        }
      });
      logger.info('Series deleted from Sonarr', { id, deleteFiles });
    } catch (error) {
      logger.error('Failed to delete series from Sonarr', { id, error });
      throw error;
    }
  }

  /**
   * Get root folders
   */
  async getRootFolders(): Promise<Array<{ id: number; path: string; freeSpace: number }>> {
    try {
      const response = await this.client.get('/api/v3/rootfolder');
      return response.data;
    } catch (error) {
      logger.error('Failed to get root folders from Sonarr', { error });
      throw error;
    }
  }

  /**
   * Get quality profiles
   */
  async getQualityProfiles(): Promise<Array<{ id: number; name: string }>> {
    try {
      const response = await this.client.get('/api/v3/qualityprofile');
      return response.data;
    } catch (error) {
      logger.error('Failed to get quality profiles from Sonarr', { error });
      throw error;
    }
  }

  /**
   * Get episodes for a series
   */
  async getEpisodes(seriesId: number): Promise<Array<{
    id: number;
    seriesId: number;
    seasonNumber: number;
    episodeNumber: number;
    title: string;
    monitored: boolean;
    hasFile: boolean;
  }>> {
    try {
      const response = await this.client.get('/api/v3/episode', {
        params: { seriesId }
      });
      return response.data;
    } catch (error) {
      logger.error('Failed to get episodes from Sonarr', { seriesId, error });
      throw error;
    }
  }

  /**
   * Update episode monitoring status
   */
  async updateEpisode(episodeId: number, updates: { monitored: boolean }): Promise<void> {
    try {
      // First get the episode
      const episode = await this.client.get(`/api/v3/episode/${episodeId}`);

      // Update with new values
      await this.client.put(`/api/v3/episode/${episodeId}`, {
        ...episode.data,
        ...updates
      });

      logger.info('Episode updated in Sonarr', { episodeId, updates });
    } catch (error) {
      logger.error('Failed to update episode in Sonarr', { episodeId, error });
      throw error;
    }
  }

  /**
   * Search for a specific episode
   */
  async searchEpisode(episodeId: number): Promise<void> {
    try {
      await this.client.post('/api/v3/command', {
        name: 'EpisodeSearch',
        episodeIds: [episodeId]
      });
      logger.info('Episode search triggered in Sonarr', { episodeId });
    } catch (error) {
      logger.error('Failed to trigger episode search in Sonarr', { episodeId, error });
      throw error;
    }
  }

  /**
   * Search for all episodes in a season
   */
  async searchSeason(seriesId: number, seasonNumber: number): Promise<void> {
    try {
      await this.client.post('/api/v3/command', {
        name: 'SeasonSearch',
        seriesId,
        seasonNumber
      });
      logger.info('Season search triggered in Sonarr', { seriesId, seasonNumber });
    } catch (error) {
      logger.error('Failed to trigger season search in Sonarr', { seriesId, seasonNumber, error });
      throw error;
    }
  }
}

File: lanflix-server/app/Infrastructure/Services/ExternalApis/SonarrClient.cs
using System.Net.Http.Json;
using System.Text.Json;
using Lanflix.Application.Common.Interfaces;
using Lanflix.Application.Common.Models;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Logging;

namespace Lanflix.Infrastructure.Services.ExternalApis;

public class SonarrClient : ISonarrClient
{
    private readonly HttpClient _httpClient;
    private readonly ILogger<SonarrClient> _logger;
    private readonly IServiceProvider _serviceProvider;
    private readonly JsonSerializerOptions _jsonOptions;

    public SonarrClient(
        HttpClient httpClient,
        IServiceProvider serviceProvider,
        ILogger<SonarrClient> logger)
    {
        _httpClient = httpClient;
        _serviceProvider = serviceProvider;
        _logger = logger;

        _jsonOptions = new JsonSerializerOptions
        {
            PropertyNameCaseInsensitive = true,
            PropertyNamingPolicy = JsonNamingPolicy.CamelCase
        };
    }

    private async Task<(string Url, string ApiKey)> GetConfigAsync(CancellationToken cancellationToken = default)
    {
        using var scope = _serviceProvider.CreateScope();
        var settingsService = scope.ServiceProvider.GetRequiredService<ISettingsService>();
        var settings = await settingsService.GetSettingsAsync(cancellationToken);
        return (settings.ExternalApis.Sonarr.Url, settings.ExternalApis.Sonarr.ApiKey);
    }

    private async Task<HttpRequestMessage> CreateRequestAsync(
        HttpMethod method,
        string requestUri,
        CancellationToken cancellationToken = default)
    {
        var (url, apiKey) = await GetConfigAsync(cancellationToken);
        
        if (string.IsNullOrEmpty(url))
        {
            throw new InvalidOperationException("Sonarr URL is not configured. Please configure Sonarr in the admin settings.");
        }

        // Ensure URL doesn't end with slash and requestUri starts with slash
        var baseUrl = url.TrimEnd('/');
        var path = requestUri.StartsWith('/') ? requestUri : '/' + requestUri;
        var fullUrl = baseUrl + path;
        
        var request = new HttpRequestMessage(method, fullUrl);
        
        if (!string.IsNullOrEmpty(apiKey))
        {
            request.Headers.Add("X-Api-Key", apiKey);
        }
        
        return request;
    }

    public async Task<bool> TestConnectionAsync(CancellationToken cancellationToken = default)
    {
        try
        {
            var request = await CreateRequestAsync(HttpMethod.Get, "/api/v3/system/status", cancellationToken);
            var response = await _httpClient.SendAsync(request, cancellationToken);
            return response.IsSuccessStatusCode;
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Failed to connect to Sonarr");
            return false;
        }
    }

    public async Task<List<SonarrSearchResult>> SearchSeriesAsync(string query, CancellationToken cancellationToken = default)
    {
        try
        {
            var request = await CreateRequestAsync(HttpMethod.Get, $"/api/v3/series/lookup?term={Uri.EscapeDataString(query)}", cancellationToken);
            var response = await _httpClient.SendAsync(request, cancellationToken);
            response.EnsureSuccessStatusCode();
            
            var results = await response.Content.ReadFromJsonAsync<List<SonarrSearchResult>>(_jsonOptions, cancellationToken);
            return results ?? new List<SonarrSearchResult>();
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Failed to search series in Sonarr: {Query}", query);
            throw;
        }
    }

    public async Task<SonarrSeries> AddSeriesAsync(AddSonarrSeriesRequest request, CancellationToken cancellationToken = default)
    {
        try
        {
            // Get subtitle settings from server settings
            using var scope = _serviceProvider.CreateScope();
            var settingsService = scope.ServiceProvider.GetRequiredService<ISettingsService>();
            var settings = await settingsService.GetSettingsAsync(cancellationToken);
            
            var payload = new
            {
                tvdbId = request.TvdbId,
                title = request.Title,
                qualityProfileId = request.QualityProfileId,
                titleSlug = request.Title.ToLower().Replace(" ", "-"),
                images = new object[] { },
                seasons = new object[] { },
                path = $"{request.RootFolderPath}/{request.Title}",
                rootFolderPath = request.RootFolderPath,
                seasonFolder = true,
                monitored = request.Monitored,
                addOptions = new
                {
                    searchForMissingEpisodes = request.SearchForMissingEpisodes,
                    // Enable subtitle search if auto-download is enabled
                    searchForSubtitles = settings.ExternalApis.Subtitles.AutoDownload
                }
            };

            var httpRequest = await CreateRequestAsync(HttpMethod.Post, "/api/v3/series", cancellationToken);
            httpRequest.Content = JsonContent.Create(payload, options: _jsonOptions);
            
            var response = await _httpClient.SendAsync(httpRequest, cancellationToken);
            response.EnsureSuccessStatusCode();

            var series = await response.Content.ReadFromJsonAsync<SonarrSeries>(_jsonOptions, cancellationToken);
            _logger.LogInformation("Series added to Sonarr: {Title} (Subtitle auto-download: {SubtitleEnabled})", 
                request.Title, settings.ExternalApis.Subtitles.AutoDownload);

            return series ?? throw new InvalidOperationException("Failed to deserialize Sonarr response");
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Failed to add series to Sonarr: {Title}", request.Title);
            throw;
        }
    }

    public async Task<List<SonarrSeries>> GetSeriesAsync(CancellationToken cancellationToken = default)
    {
        try
        {
            var request = await CreateRequestAsync(HttpMethod.Get, "/api/v3/series", cancellationToken);
            var response = await _httpClient.SendAsync(request, cancellationToken);
            response.EnsureSuccessStatusCode();
            
            var series = await response.Content.ReadFromJsonAsync<List<SonarrSeries>>(_jsonOptions, cancellationToken);
            return series ?? new List<SonarrSeries>();
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Failed to get series from Sonarr");
            throw;
        }
    }

    public async Task<SonarrSeries?> GetSeriesByTvdbIdAsync(int tvdbId, CancellationToken cancellationToken = default)
    {
        try
        {
            var series = await GetSeriesAsync(cancellationToken);
            return series.FirstOrDefault(s => s.TvdbId == tvdbId);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Failed to get series by TVDB ID: {TvdbId}", tvdbId);
            throw;
        }
    }

    public async Task<SonarrQueueResponse> GetQueueAsync(int page = 1, int pageSize = 20, CancellationToken cancellationToken = default)
    {
        try
        {
            var request = await CreateRequestAsync(HttpMethod.Get, $"/api/v3/queue?page={page}&pageSize={pageSize}", cancellationToken);
            var response = await _httpClient.SendAsync(request, cancellationToken);
            response.EnsureSuccessStatusCode();
            
            var queueResponse = await response.Content.ReadFromJsonAsync<SonarrQueueResponse>(_jsonOptions, cancellationToken);
            return queueResponse ?? new SonarrQueueResponse();
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Failed to get queue from Sonarr");
            throw;
        }
    }

    public async Task DeleteSeriesAsync(int id, bool deleteFiles = false, CancellationToken cancellationToken = default)
    {
        try
        {
            var request = await CreateRequestAsync(HttpMethod.Delete, $"/api/v3/series/{id}?deleteFiles={deleteFiles}", cancellationToken);
            var response = await _httpClient.SendAsync(request, cancellationToken);
            response.EnsureSuccessStatusCode();
            
            _logger.LogInformation("Series deleted from Sonarr: {Id}", id);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Failed to delete series from Sonarr: {Id}", id);
            throw;
        }
    }

    public async Task<List<SonarrRootFolder>> GetRootFoldersAsync(CancellationToken cancellationToken = default)
    {
        try
        {
            var request = await CreateRequestAsync(HttpMethod.Get, "/api/v3/rootfolder", cancellationToken);
            var response = await _httpClient.SendAsync(request, cancellationToken);
            response.EnsureSuccessStatusCode();
            
            var folders = await response.Content.ReadFromJsonAsync<List<SonarrRootFolder>>(_jsonOptions, cancellationToken);
            return folders ?? new List<SonarrRootFolder>();
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Failed to get root folders from Sonarr");
            throw;
        }
    }

    public async Task<List<SonarrQualityProfile>> GetQualityProfilesAsync(CancellationToken cancellationToken = default)
    {
        try
        {
            var request = await CreateRequestAsync(HttpMethod.Get, "/api/v3/qualityprofile", cancellationToken);
            var response = await _httpClient.SendAsync(request, cancellationToken);
            response.EnsureSuccessStatusCode();
            
            var profiles = await response.Content.ReadFromJsonAsync<List<SonarrQualityProfile>>(_jsonOptions, cancellationToken);
            return profiles ?? new List<SonarrQualityProfile>();
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Failed to get quality profiles from Sonarr");
            throw;
        }
    }

    public async Task<List<SonarrEpisode>> GetEpisodesAsync(int seriesId, CancellationToken cancellationToken = default)
    {
        try
        {
            var request = await CreateRequestAsync(HttpMethod.Get, $"/api/v3/episode?seriesId={seriesId}", cancellationToken);
            var response = await _httpClient.SendAsync(request, cancellationToken);
            response.EnsureSuccessStatusCode();
            
            var episodes = await response.Content.ReadFromJsonAsync<List<SonarrEpisode>>(_jsonOptions, cancellationToken);
            return episodes ?? new List<SonarrEpisode>();
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Failed to get episodes from Sonarr: {SeriesId}", seriesId);
            throw;
        }
    }

    public async Task SearchEpisodeAsync(int episodeId, CancellationToken cancellationToken = default)
    {
        try
        {
            var payload = new
            {
                name = "EpisodeSearch",
                episodeIds = new[] { episodeId }
            };

            var httpRequest = await CreateRequestAsync(HttpMethod.Post, "/api/v3/command", cancellationToken);
            httpRequest.Content = JsonContent.Create(payload, options: _jsonOptions);
            
            var response = await _httpClient.SendAsync(httpRequest, cancellationToken);
            response.EnsureSuccessStatusCode();

            _logger.LogInformation("Episode search triggered in Sonarr: {EpisodeId}", episodeId);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Failed to trigger episode search in Sonarr: {EpisodeId}", episodeId);
            throw;
        }
    }

    public async Task SearchSeasonAsync(int seriesId, int seasonNumber, CancellationToken cancellationToken = default)
    {
        try
        {
            var payload = new
            {
                name = "SeasonSearch",
                seriesId,
                seasonNumber
            };

            var httpRequest = await CreateRequestAsync(HttpMethod.Post, "/api/v3/command", cancellationToken);
            httpRequest.Content = JsonContent.Create(payload, options: _jsonOptions);
            
            var response = await _httpClient.SendAsync(httpRequest, cancellationToken);
            response.EnsureSuccessStatusCode();

            _logger.LogInformation("Season search triggered in Sonarr: Series={SeriesId}, Season={SeasonNumber}", seriesId, seasonNumber);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Failed to trigger season search in Sonarr: Series={SeriesId}, Season={SeasonNumber}", seriesId, seasonNumber);
            throw;
        }
    }
}
